build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -702,6 +702,48 @@ message NodeProperty {
     string value = 2;
 }
 
+// Node properties for [FileNodes][build.bazel.remote.execution.v2.FileNode],
+// [DirectoryNodes][build.bazel.remote.execution.v2.DirectoryNode], and
+// [SymlinkNodes][build.bazel.remote.execution.v2.SymlinkNode]. The server is
+// responsible for specifying the properties that it accepts.
+//
+message NodeProperties {
+  // A list of string-based
+  // [NodeProperties][build.bazel.remote.execution.v2.NodeProperty].
+  repeated NodeProperty properties = 1;
+
+  // The keys of node properties whose values are valid in this message.
+  // This is required as protobuf doesn't provide a way to distinguish between
```

Comment:
Am I misunderstanding the documentation?

> Note that for scalar message fields, once a message is parsed there's no way of telling whether a field was explicitly set to the default value (for example whether a boolean was set to false) or just not set at all: you should bear this in mind when defining your message types. For example, don't have a boolean that switches on some behaviour when set to false if you don't want that behaviour to also happen by default. Also note that if a scalar message field is set to its default, the value will not be serialized on the wire.

https://developers.google.com/protocol-buffers/docs/proto3#default

Reply:
Ah, I see. You can tell whether a non-scalar field (i.e., message field) is set or not. I didn't check for scalar fields. Below, that would apply to unix_mode, but not mtime.

I really don't like specifying the keys in addition to the fields - it seems that that will cancel out most of the expected size benefits. Would it be possible to always require these to be wrapped in a proto message instead, i.e., forbid scalar fields?

Reply:
It would definitely be nice to avoid the need for the extra keys message with boolean fields. I just discovered `wrappers.proto` with `Int32Value` and similar wrapper messages. That seems like a perfect fit and I'll change this branch accordingly.

The `NodePropertyKeys` message will still be needed for `output_node_properties` and capabilities but not in the directory tree.

Reply:
I've updated the branch. Does this look good to you now?